native/acter/src/api/reactions.rs

```diff
@@ -0,0 +1,218 @@
+use ruma_events::{reaction::ReactionEventContent, relation::Annotation, MessageLikeEventType};
+use std::ops::Deref;
+use tokio::sync::broadcast::Receiver;
+use tokio_stream::{wrappers::BroadcastStream, Stream};
+use tracing::trace;
+
+use super::{client::Client, RUNTIME};
+
+impl Client {
+    pub async fn wait_for_reaction(
+        &self,
+        key: String,
+        timeout: Option<Box<Duration>>,
+    ) -> Result<Reaction> {
+        let client = self.clone();
+        RUNTIME
+            .spawn(async move {
+                let AnyActerModel::Reaction(reaction) = client.wait_for(key.clone(), timeout).await? else {
+                    bail!("{key} is not a reaction");
+                };
+                let room = client
+                    .core
+                    .client()
+                    .get_room(&reaction.meta.room_id)
+                    .context("Room not found")?;
+                Ok(Reaction {
+                    client: client.clone(),
+                    room,
+                    inner: reaction,
+                })
+            })
+            .await?
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Reaction {
+    client: Client,
+    room: Room,
+    inner: models::Reaction,
+}
+
+impl Deref for Reaction {
+    type Target = models::Reaction;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl Reaction {
+    pub fn event_id_str(&self) -> String {
+        self.inner.event_id().to_string()
+    }
+
+    pub fn sender(&self) -> OwnedUserId {
+        self.inner.meta.sender.clone()
+    }
+
+    pub fn origin_server_ts(&self) -> u64 {
+        self.inner.meta.origin_server_ts.get().into()
+    }
+
+    pub fn relates_to(&self) -> String {
+        self.inner.relates_to.event_id.to_string()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ReactionManager {
+    client: Client,
+    room: Room,
+    inner: models::ReactionManager,
+}
+
+impl Deref for ReactionManager {
+    type Target = models::ReactionManager;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl ReactionManager {
+    pub(crate) fn new(
+        client: Client,
+        room: Room,
+        inner: models::ReactionManager,
+    ) -> ReactionManager {
+        ReactionManager {
+            client,
+            room,
+            inner,
+        }
+    }
+
+    pub async fn send_reaction(&self, event_id: String, key: String) -> Result<OwnedEventId> {
+        let room = self.room.clone();
+        let client = room.client();
+        let my_id = client.user_id().context("User not found")?.to_owned();
+        let event_id = EventId::parse(&event_id)?;
+
+        RUNTIME
+            .spawn(async move {
+                let member = room
+                    .get_member(&my_id)
+                    .await?
+                    .context("Couldn't find me among room members")?;
+                if !member.can_send_message(MessageLikeEventType::RoomMessage) {
```

Comment:
You sure this is the right check? Shouldn't we check for Annotations? I think they have a different permission-level ...

Reply:
I thought it is correct to check `MessageLikeEventType::RoomMessage` because `room.send()` is called below that.
Anyway fixed [here](https://github.com/acterglobal/a3/pull/1208/commits/ec6cbeba3c242b3aabfdb3475e3b8b9d0279a988).